File: c++-projects/ch03/money.cpp

#include <iostream>
#include <cctype>
#include <cmath>
#include <stdexcept>
#include "money.h"

static std::string& titlecase(std::string& str);

Money::Money(int unit, double fraction, std::string name):
unit(unit), fraction(fraction), name(titlecase(name)){}

const std::string& Money::get_name() const
{
    return name;
}

int Money::get_unit() const
{
    return unit;
}
double Money::get_fraction() const
{
    return fraction;
}

bool Money::operator==(const Money& m)
{
    return name == m.name;
}

const Money Money::operator+(const Money& m) const
{
    if (name == m.name)
    {
        return Money(unit + m.unit, fraction + m.fraction);
    }
    throw std::runtime_error("Can't add two different Money type");
}

const Money Money::operator-(const Money& m) const
{
    if (name == m.name)
    {
        return Money(unit - m.unit, fraction - m.fraction);
    }
    throw std::runtime_error("Can't subtract two different Money type");
}

std::ostream& operator<<(std::ostream& os, const Money& m)
{
    if (m.unit != 0)
        os << m.unit << "."<< std::abs(m.fraction) << " "<< m.name;
    else
        os << m.unit << "."<< m.fraction << " "<< m.name;
    return os;
}

static std::string& titlecase(std::string& str)
{
    // if the string is empty
    if (str.size() == 0) return str;
    *str.begin() = std::toupper(*str.begin());
    for(auto s = str.begin()+1, e = str.end(); s != e; ++s)
        *s = std::tolower(*s);
    return str;
}

File: c++-projects/ch03/money.py
#!/usr/bin/env python3

class Money:
    '''class declare money type'''
    def __init__(self, unit: int, fraction: float, name: str = 'Naira'):
        'initialize money class'
        self.__unit = unit
        self.__fraction = fraction
        self.__name = name.casefold()

    def get_name(self) -> str:
        'get the money name'
        return self.__name

    def get_unit(self) -> int:
        'get the money type unit'
        return self.__unit

    def get_fraction(self) -> float:
        'get money type fractional type'
        return self.__fraction

    def __str__(self) -> str:
        '''string repr of the money data type'''
        if self.__unit != 0:
            return f"{self.__unit}.{abs(self.__fraction)} {self.__name}"
        return self.__repr__()

    def __repr__(self) -> str:
        '''string repr for money type'''
        return f"{self.__unit}.{self.__fraction} {self.__name}"

    def __add__(self, m):
        '''add two same type of money'''
        if self.__eq__(m):
            return Money(self.__unit + m.__unit,
                self.__fraction + m.__fraction)
        return Exception("Can't add two different Money type")

    def __sub__(self, m):
        'minus type of money'
        #self, m = self.__sorter(m) # superflulious here
        if self.__same_type(m):
            return Money(self.__unit - m.get_unit(),
                self.__fraction - m.get_fraction())
        return Exception("Can't subtract two different Money type")

    def __eq__(self, m, /) -> bool:
        return self.__name == m.get_name()

    def __lt__(self, m):
        'implement less than'
        return ((self.__unit < m.get_unit()) and (self.__fraction < m.get_fraction()))

    def __gt__(self, m):
        'implement greater than'
        return ((self.__unit > m.get_unit()) and (self.__fraction > m.get_fraction()))

    def __sorter(self, m):
        'sort the type of money in descending order'
        if self > m:
            return self, m
        return m, self

    def __same_type(self, m) -> bool:
        'check type of Money'
        return self.__eq__(m)
